pkg/volume/csi/csi_attacher.go

```diff
@@ -387,13 +368,23 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 		return volumetypes.NewTransientOperationFailure(log("attacher.MountDevice failed to determine if the node service has VOLUME_MOUNT_GROUP capability: %v", err))
 	}
 
+	err = saveVolumeData(dataDir, volDataFileName, data)
+	if err != nil {
+		errMsg := log("failed to save volume info data: %v", err)
+		klog.Error(errMsg)
+		if err := removeMountDir(c.plugin, deviceMountPath); err != nil {
+			klog.Error(log("attacher.MountDevice failed to remove mount dir after error [%s]: %v", deviceMountPath, err))
+		}
+		return errors.New(errMsg)
+	}
+
 	if driverSupportsCSIVolumeMountGroup {
 		klog.V(3).Infof("Driver %s supports applying FSGroup (has VOLUME_MOUNT_GROUP node capability). Delegating FSGroup application to the driver through NodeStageVolume.", csiSource.Driver)
 		nodeStageFSGroupArg = deviceMounterArgs.FsGroup
 	}
 
 	fsType := csiSource.FSType
-	err = csi.NodeStageVolume(ctx,
+	csiRPCError := csi.NodeStageVolume(ctx,
```

Comment:
As the defer is removed, We might be able to re-use the err variable. csi client wrapper the original error from the server, so the error might not be a PRC error.

Reply:
so, to make it clear that we are are doing something special in case of CSI errors, I would like to keep different variable name for CSI errors. This will also ensure that a future modification to the code doesn't accidentally trample the error variable.